pauliopt/gates.py

```diff
@@ -112,18 +116,158 @@ def gadgets(self):
     def to_qiskit(self):
         pass
 
+    @abstractmethod
+    def inverse(self):
+        pass
+
+    def apply_permutation(self, permutation: list) -> None:
+        register = list(range(len(permutation)))
+        self.qubits = tuple(
+            [permutation[register.index(qubit)] for qubit in self.qubits]
+        )
 
-class PhaseGate(Gate):
-    def __init__(self, phase, *qubits):
+    def copy(self):
+        return self.__class__(*self.qubits)
+
+
+class PhaseGate(Gate, ABC):
+    def __init__(self, phase: Angle, *qubits):
         super().__init__(*qubits)
         self.phase = phase
 
     def __repr__(self) -> str:
         args = map(repr, self.qubits)
         return f"{self.name}({self.phase}, {', '.join(args)})"
 
+    def get_phase_as_float(self):
+        return self.phase if not isinstance(self.phase, Angle) else self.phase.to_qiskit
+
+    def get_phase_as_angle(self):
+        return self.phase if isinstance(self.phase, Angle) else Angle(self.phase)
+
+    def inverse(self):
+        return self.__class__(-self.phase, *self.qubits)
+
+    def copy(self):
+        return self.__class__(self.phase, self.qubits)
+
+
+PROPAGATION_H = {
+    "X": (Pauli.Z, 1),
+    "Y": (Pauli.Y, -1),
+    "Z": (Pauli.X, 1),
+    "I": (Pauli.I, 1),
+}
+
+PROPAGATION_S = {
+    "X": (Pauli.Y, -1),
+    "Y": (Pauli.X, 1),
+    "Z": (Pauli.Z, 1),
+    "I": (Pauli.I, 1),
+}
+
+PROPAGATION_CX = {
+    "XX": (Pauli.X, Pauli.I, 1),
+    "XY": (Pauli.Y, Pauli.Z, 1),
+    "XZ": (Pauli.Y, Pauli.Y, -1),
+    "XI": (Pauli.X, Pauli.X, 1),
+    "YX": (Pauli.Y, Pauli.I, 1),
+    "YY": (Pauli.X, Pauli.Z, -1),
+    "YZ": (Pauli.X, Pauli.Y, 1),
+    "YI": (Pauli.Y, Pauli.X, 1),
+    "ZX": (Pauli.Z, Pauli.X, 1),
+    "ZY": (Pauli.I, Pauli.Y, 1),
+    "ZZ": (Pauli.I, Pauli.Z, 1),
+    "ZI": (Pauli.Z, Pauli.I, 1),
+    "IX": (Pauli.I, Pauli.X, 1),
+    "IY": (Pauli.Z, Pauli.Y, 1),
+    "IZ": (Pauli.Z, Pauli.Z, 1),
+    "II": (Pauli.I, Pauli.I, 1),
+}
+
+
+class CliffordGate(Gate, ABC):
+    def __init__(self, *qubits):
+        super().__init__(*qubits)
+
+    @abstractmethod
+    def get_h_s_cx_decomposition(self) -> List[Union["H", "S", "CX"]]:
+        """
+        Every clifford must me decomposable into a list of H, S and CX gates.
+        Returns:
+
+        """
+        pass
+
+    def propagate_pauli(self, gadget: "pauliopt.pauli.pauli_gadget.PauliGadget"):
+        """
+        Propagate a pauli gate through a gadget using the H, S, CX decomposition rules.
+
+        One can define for H, S and CX propagation rules, which are defined in the dictionaries above.
+        Args:
+            gadget:
+
+        Returns:
+
+        """
+        h_s_cx_decomposition = self.get_h_s_cx_decomposition()
+        for gate in reversed(h_s_cx_decomposition):
+            if gate.name == "H":
```

Comment:
the logic of these 3 cases seem quite similar. Is it possible to combine it into one by having a mapping from gate.name to propagation dictionary?

Reply:
and a loop to update the Paulis?

Reply:
What exactly do you mean here loop over the qubits? 
But of a CX a control is kind of dependent of the target pauli and vice versa.